### PR TITLE
Implement `into` modifier on parameter types

### PR DESCRIPTION
**Pull request by odersky**
in the thread https://contributors.scala-lang.org/t/proposed-changes-and-restrictions-for-implicit-conversions/4923
we discussed two changes that would make implicit conversions largely redundant. One proposed change was implemented in #14497. The other is implemented here. It's based on #14497.

The idea of this PR is to have some way to specify that a method parameter accepts implicit conversions on its arguments. Then the feature warning on implicit conversion use would be suppressed in this case. In the previous thread I proposed
`~` to mark convertible arguments but I now feel this is too cryptic. Instead there is a `into` prefix in the 
parameter type. E.g.
```scala
  def f(x: into T) = ...
```
For Scala 2, we introduce an annotation on the parameter that expresses the same thing:
```scala
  // proposed Scala-2 syntax:
  def f(@allowConversions x: T) = 
```
A larger example:
```scala
given Conversion[String, Text] = Text(_)

@main def Test =

  def f(x: into Text, y: => into Text, zs: into Text*) =
    println(s"${x.str} ${y.str} ${zs.map(_.str).mkString(" ")}")

  f("abc", "def")  // ok, no feature warning
  f("abc", "def", "xyz", "uvw")  // also ok
  f("abc", "def", "xyz", Text("uvw"))  // also ok
```
A larger example is a parser combinator library that works without unrestricted implicit conversions:
[tests/run/Parser.scala](https://github.com/lampepfl/dotty/blob/0f113da1bacc01603740cde4a5bafbfae39895f0/tests/run/Parser.scala)

### CODE REVIEW

**Comment by nicolasstucki** [comment]
Will we need to extend this concept at some point to lambdas?

**Comment by odersky** [comment]
I went ahead and made this an experimental feature so that people can try it out. 

Points to note:

 - When Scala 3 conversions are applied implicitly they currently require an `implicitConversions` language import _at the use site_. (Old style Scala 2 implicit conversions required the language import at definition site, but that turned out to be toothless). 
 - With this PR there is now an exception: No language import is required if the implicit conversion is on a method argument where the corresponding parameter is marked `into`. 

This technique addresses the situations where implicit conversions were found to be essential and hard to replace in the discussion thread cited above.

**Comment by deusaquilus** [comment]
@odersky I'm so sorry to keep pestering you about this but after you closed #16002 I really don't know what else to do. For Quill to work I need to have both implicit conversions `Quoted[T] => T` and `T => Quoted[T]`. The `into` capability might help with the `T => Quoted[T]` part but it's the `Quoted[T] => T` that is much more important (I mentioned that on the original thread [here](https://contributors.scala-lang.org/t/proposed-changes-and-restrictions-for-implicit-conversions/4923/167?u=deusaquilus)).

At this point, the only thing that can save Quill from an eternity of needing implicit-conversions is some kind of injectable `Quoted[T] => T` conversion that would be injected into `quote { ... }` blocks. Without something like that, every other kind of feature like this one is completely pointless for Quill.

**Comment by odersky** [comment]
@deusaquilus I don't know a solution either. Maybe Quill won't profit from `into`. I am not sure what else can be done. For the moment, the language import might be the only way to do it.

**Comment by odersky** [comment]
@dwijnand Can you give this a review?

**Comment by odersky** [comment]
The reason to merge this now is that I'd like to experiment with the compiler to see whether we can get rid of implicit conversions. We need to make progress on this since old style implicit conversions will be deprecated at some point and new style implicit conversions require a language import at the use site. 

To make this work, we need to have alternative mechanisms in place that make implicit conversions necessary only in exceptional cases. `into` is such a mechanism, and right now it is not clear whether it is already sufficient or we need something else as well. Now, the compiler contains thousands of occurrences of implicit conversions. So it is an ideal testing ground to see how far we get in getting rid of them. For before we start we need `into`.

File: docs/_docs/reference/experimental/into-modifier.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+```scala
+def concatAll(xss: into IterableOnce[Char]*): List[Char] =
+  xss.foldLeft(List[Char]())(_ ++ _)
+```
+Here, the call
+```scala
+concatAll(List('a'), "bc", Array('d', 'e'))
+```
+would apply two _different_ implicit conversions: the conversion from `String` to `List[Char]` gets applied to the second argument and the conversion from `Array[Char]` to `Iterable[Char]` gets applied to the third argument.
```
Comment (bishabosha) [inline docs/_docs/reference/experimental/into-modifier.md:59]:
```suggestion
would apply two _different_ implicit conversions: the conversion from `String` to `Iterable[Char]` gets applied to the second argument and the conversion from `Array[Char]` to `Iterable[Char]` gets applied to the third argument.
```

File: docs/_docs/reference/experimental/into-modifier.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+scala> val xs = List(1, 2, 3)
+scala> xs.flatMap(x => x.toString * x)
+val res2: List[Char] = List(1, 2, 2, 3, 3, 3)
+```
+Here, the conversion from `String` to `Iterable[Char]` is applied on the results of `flatMap`'s function argument when it is applied to the elements of `xs`.
+
+## Vararg arguments
+
+When applied to a vararg parameter, `into` allows a conversion on each argument value individually. For example, consider a method `concatAll` that concatenates a variable
+number of `Seq[Char]` arguments, and also allows implicit conversions into `Seq[Char`:
```
Comment (bishabosha) [inline docs/_docs/reference/experimental/into-modifier.md:49]:
```suggestion
number of `IterableOnce[Char]` arguments, and also allows implicit conversions into `IterableOnce[Char]`:
```

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 3]
```diff
@@ -3948,27 +3957,48 @@ object Types {
...
+      def wrapConvertible(tp: Type) =
+        AppliedType(defn.IntoType.typeRef, tp :: Nil)
+
+      /** Add `Into[..] to the type itself and if it is a function type, to all its
+       *  curried result type(s) as well.
+       */
+      def addInto(tp: Type): Type = tp match
+        case tp @ AppliedType(tycon, args) if tycon.typeSymbol == defn.RepeatedParamClass =>
+          tp.derivedAppliedType(tycon, addInto(args.head) :: Nil)
+        case tp @ AppliedType(tycon, args) if defn.isFunctionType(tp) =>
```
Comment (KacperFKorban) [inline compiler/src/dotty/tools/dotc/core/Types.scala:3977]: Will this handle polyfunctions?

File: library/src/scala/runtime/stdLibPatches/language.scala [hunk 4]
```diff
@@ -74,6 +74,14 @@ object language:
      */
     @compileTimeOnly("`captureChecking` can only be used at compile time in import statements")
     object captureChecking
+
+    /** Experimental support for automatic conversions of arguments, without requiring
+     *  a langauge import `import scala.language.implicitConversions`.
+     *
+     *  @see [[https://dotty.epfl.ch/docs/reference/experimental/cc]]
```
Comment (KacperFKorban) [inline library/src/scala/runtime/stdLibPatches/language.scala:81]: `https://dotty.epfl.ch/docs/reference/experimental/into-modifier`

**Review by KacperFKorban — APPROVED** [review]
LGTM overall. Just added some minor comments.

Also sorry for causing this PR to hang for so long, I must've dismissed the notification by accident.

**Comment by odersky** [comment]
@KacperFKorban Thanks for the review!

**Pull request merged**